### PR TITLE
Client and shadow traffic unit tests

### PR DESCRIPTION
**Pull request by scottmcmaster**

### CODE REVIEW

File: promoted_python_delivery_client/client/client.py [hunk 1]
```diff
@@ -108,10 +116,13 @@ def _log_to_metrics(self,
                         cohort_membership: Optional[CohortMembership],
                         exec_svr: ExecutionServer) -> None:
         log_request = self._create_log_request(delivery_request, response, cohort_membership, exec_svr)
-        try:
-            self.api_metrics.run_metrics_logging(log_request)
-        except Exception as ex:
-            logging.error("Error logging to metrics", exc_info=ex)
+        if self.only_log_metrics_request:
```
Comment (prm-dan) [inline promoted_python_delivery_client/client/client.py]: I assumed `only_log_metrics_request` would have logged to the Metrics API log method.  Maybe rename this field to "text_log_metrics_request".

File: promoted_python_delivery_client/client/client.py [hunk 2]
```diff
@@ -159,6 +170,8 @@ def _deliver_shadow_traffic(self, delivery_request: DeliveryRequest) -> None:
             logging.error("Error sending shadow traffic", exc_info=ex)
 
     def _should_apply_treatment(self, cohort_membership: Optional[CohortMembership] = None) -> bool:
+        if self.apply_treatment_checker is not None:
+            return self.apply_treatment_checker(cohort_membership)
         if cohort_membership is None:
```
Comment (prm-dan) [inline promoted_python_delivery_client/client/client.py:175]: Do you want the remaining implementation to be a default implementation of apply_treatment_checker?

File: promoted_python_delivery_client/client/client.py [hunk 3]
```diff
@@ -50,12 +50,16 @@ def __init__(self,
                  metrics_timeout_millis: int = DEFAULT_METRICS_TIMEOUT_MILLIS,
                  max_request_insertions: int = DEFAULT_MAX_REQUEST_INSERTIONS,
                  shadow_traffic_delivery_rate: float = 0,
-                 perform_checks: bool = False) -> None:
+                 perform_checks: bool = False,
+                 only_log_metrics_request: bool = False,
+                 apply_treatment_checker: Optional[Callable[[Optional[CohortMembership]], bool]] = None):
```
Comment (prm-dan) [inline promoted_python_delivery_client/client/client.py:55]: Is this for mocking in tests?

**Review by prm-dan — APPROVED** [review]

File: promoted_python_delivery_client/client/client.py (hunk 3 above)
Comment (scottmcmaster) [reply promoted_python_delivery_client/client/client.py:55]: It is a feature in the other SDKs.

File: promoted_python_delivery_client/client/client.py (hunk 2 above)
Comment (scottmcmaster) [reply promoted_python_delivery_client/client/client.py:175]: I'd rather not, as it seems like that would only unlock the possibility that it be nulled out, which would then have to be checked for at the call site.

File: promoted_python_delivery_client/client/client.py (hunk 1 above)
Comment (scottmcmaster) [reply promoted_python_delivery_client/client/client.py]: Unfortunately "log" is overloaded I guess. I will find some different name.

**Pull request merged**